rfcs/008-map-indexes.md

```diff
@@ -0,0 +1,181 @@
+
+`{<database>, ?VIEWS, <view_signature>, ?VIEW_ID_INDEX, <_id>, <view_id>} -> [emitted keys]`
+
+Each index will be built and updated via the Background job queue [RFC Link TBD]. When a request for a view is received, the request process will add a job item onto the background queue for the index to be updated. A worker will take the item off the queue and update the index. Once the index has been built, the request will return with the results. This process can also be optimised in two ways. Firstly, using a new couch_events system to listen for document changes in an database and then adding indexing jobs to the queue to keep indexes warm. The second optimisation is if the index only requires a small update, rather update the index in the http request process instead of doing the work via the background queue.
+
+Initially the building of an index will be a single worker running through the changes feed and creating the index. Ideally it would be nice to parallelise that work so that multiple workers could build the index at the same time. This will reduce build times. This can be done by fetching the boundary keys for the changes feed, splitting those key ranges amongst different workers to build different parts of the index. This will require that for each document update processed, the worker must check the revision key space to determine if the document is the latest revision of the document. If it is not it should discard it. The other requirement is that CouchDB could only start serving the index once the update sequence is up to date AND all the workers have completed building their section of the index.
+
+### View clean up
+
+When a design document is changed, new indexes will be built and grouped under a new `View Signature`. The old map indexes will still be in FDB.  CouchDB will need to monitor `View Signatures` and be able to remove old indexes. To do this we create the following data model:
+
+`(<database>, ?VIEW_DDOC_IDS, <design_doc_id>) = ViewSignature`
+`(<database>, ?VIEW_SIGS, <view_signature>) = Counter`
+
+When a design document is created, CouchDB will store the design document id and the view signature and set the view signature counter to one. On update or deletion of a design document, CouchDB will get the old signature and decrement the counter. 
+If the counter is 0, then CouchDB will remove the old view index.
+
+### Stale = “ok” and stable = true
+
+ With the consistency guarantee’s CouchDB will get from FDB, `stable = true` will no longer be an option that CouchDB would support and so the argument would be ignored. Similar `stale = “ok”` would now be translated to `update = false`.
```

Comment:
Out of scope, but as an aside it'd be excellent to offer a "bounded staleness" option; i.e., "serve the view stale if it's missing no more than <N> updates, or is lagging by no more than <M> seconds". We don't have the bookkeeping to offer that at the moment, just planting the idea here in case anyone thinks of an elegant way to track that.